applications/nrf5340_audio/src/modules/power_meas.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(power_meas, CONFIG_LOG_POWER_MODULE_LEVEL);
+
+/** Number of power rails */
+#define POWER_RAIL_NUM 4U
+
+/** Power rail configuration */
+struct power_rail_config {
+	/** Name */
+	const char *name;
+	/** INA231 sensor device */
+	const struct device *sensor;
+};
+
+/** Configuration for all power rails */
+static const struct power_rail_config rail_config[POWER_RAIL_NUM] = {
+	{"VBAT", DEVICE_DT_GET(DT_NODELABEL(vbat_sensor))},
+	{"VDD1_CODEC", DEVICE_DT_GET(DT_NODELABEL(vdd1_codec_sensor))},
+	{"VDD2_CODEC", DEVICE_DT_GET(DT_NODELABEL(vdd2_codec_sensor))},
+	{"VDD2_NRF", DEVICE_DT_GET(DT_NODELABEL(vdd2_nrf_sensor))},
+};
+/** Enabled rails (same index as rail_config) */
+static bool rail_enabled[POWER_RAIL_NUM];
+
+/** Measurement thread stack */
+K_THREAD_STACK_DEFINE(meas_stack, CONFIG_POWER_MEAS_STACK_SIZE);
+/** Measurement thread */
+static struct k_thread meas_thread;
+/** Measurement signal semaphore */
+K_SEM_DEFINE(meas_sem, 0, 1);
+/** Measurement enabled flag */
+static bool meas_enabled;
+
+/**
+ * @brief Read a power rail data (voltage/current/power) and log results.
+ *
+ * @param[in] config Power rail configuration.
+ *
+ * @retval 0 on success
+ * @retval -errno on sensor fetch/get error.
+ */
+static int read_and_log(const struct power_rail_config *config)
+{
+	int ret;
+	struct sensor_value voltage, current, power;
+
+	ret = sensor_sample_fetch(config->sensor);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = sensor_channel_get(config->sensor, SENSOR_CHAN_VOLTAGE, &voltage);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = sensor_channel_get(config->sensor, SENSOR_CHAN_CURRENT, &current);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = sensor_channel_get(config->sensor, SENSOR_CHAN_POWER, &power);
+	if (ret < 0) {
+		return ret;
+	}
+
+	LOG_INF("%-10s:\t%.3fV, %06.3fmA, %06.3fmW", config->name,
+		sensor_value_to_double(&voltage),
+		sensor_value_to_double(&current) * 1000.0f,
+		sensor_value_to_double(&power) * 1000.0f);
+
+	return 0;
+}
+
+/** @brief Measurement thread */
+static void meas_thread_fn(void *dummy1, void *dummy2, void *dummy3)
+{
+	ARG_UNUSED(dummy1);
+	ARG_UNUSED(dummy2);
+	ARG_UNUSED(dummy3);
+
+	while (1) {
+		if (!meas_enabled) {
+			k_sem_take(&meas_sem, K_FOREVER);
+		}
+
+		for (size_t i = 0U; i < POWER_RAIL_NUM; i++) {
+			int ret;
+
+			if (!rail_enabled[i]) {
+				continue;
+			}
+
+			ret = read_and_log(&rail_config[i]);
+			if (ret < 0) {
```

Comment:
If we follow Zephyr coding guidelines, we should be using essentially boolean expressions, e.g. `ret < 0` or `ret != 0`, but not `ret`. Since error is specified only in case ret < 0, I tend to prefer that choice (> 0 is not specified as an error or a valid value in many APIs).